sklearn/impute/tests/test_impute.py

```diff
@@ -1364,3 +1364,27 @@ def test_imputation_order(order, idx_order):
                                random_state=0).fit(X)
         idx = [x.feat_idx for x in trs.imputation_sequence_]
         assert idx == idx_order
+
+
+@pytest.mark.parametrize(
+    "Imputer,arg",
+    [(SimpleImputer, 'strategy'), (IterativeImputer, 'initial_strategy')])
+@pytest.mark.parametrize(
+    "strategy",
+    ['mean', 'median', 'most_frequent', "constant"])
+def test_imputation_keep_missing_features(Imputer, arg, strategy):
+    X = np.array([[1, np.nan, 2], [3, np.nan, np.nan]])
+
+    imputer = Imputer(**{arg: strategy, 'keep_missing_features': True})
+    assert X.shape == imputer.fit_transform(X).shape
+
+    if strategy == "constant" and Imputer is IterativeImputer:
+        return
+
+    imputer = Imputer(**{arg: strategy, 'keep_missing_features': False})
+    X_trans = imputer.fit_transform(X)
+
+    if strategy == "constant":
+        assert X.shape == X_trans.shape
```

Comment:
This is a strange inconsistency. I think we should follow keep_missing_features in the constant case too and drop the feature if `keep_missing_features=False`.

Reply:
`SimpleImputer`'s `constant` strategy never dropped the features; and I don't think it should, because we can use it after the other imputers to prevent `NaN`s.

So the problem is that `IterativeImputer(strategy='constant', 'keep_missing_features'=False)` drops the features, whilst its `SimpleImputer` doesn't.

Reply:
I'd personally rather consistency across `strategy`. However, I'm happy to leave 'constant' as exceptional for now, but it should be documented.

I don't think we should be using it after other imputers to prevent NaNs. I think a guaranteed invariance for imputers should be that after transform there are no more missing values. Hence my support for @rth's `fill_missing_features` idea.

Reply:
Should there be a `fill_missing_features_value` parameter, then?